libs/database/src/lib/services/ads.service.ts

```diff
@@ -489,4 +485,48 @@ export class AdsService {
           adsPlacement: adsPlacement,
         };
   };
+
+  getAds = async (accountId: string) => {
+    //select all ads that user Balance > budget
+    const ads = await this._adsCampaignModel
+      .aggregate(pipe2AvaialableAdsCampaign())
+      .exec();
+    console.log('===--=getAds');
+    console.log(ads);
+    console.log(ads.length);
+
+    const releventScore = await this.dataService.personalizeContents(
+      accountId,
+      ads.map((a) => a.adsRef.$id || a.adsRef.oid),
+    ); // O ?
+    const cpms: AdsCpm[] = [];
+    for (let i = 0; i < ads.length; i++) {
+      cpms[i] = { cpm: ads.length * Environment.ADS_MINIMUM_CPM };
+      if (cpms[i].cpm > ads[i].budgetLeft) cpms[i].cpm = ads[i].budgetLeft;
+      cpms[i].bidding_cpm = cpms[i].cpm;
+      cpms[i].relevance_score =
+        releventScore[ads[i].adsRef.$id || ads[i].adsRef.oid];
+      cpms[i].ranking_score = cpms[i].relevance_score * cpms[i].bidding_cpm;
+      cpms[i].adsCampignId = String(ads[i]._id);
+    }
+    //
+    const sortedCpms = cpms.sort((a, b) =>
+      a.ranking_score < b.ranking_score ? -1 : 1,
+    );
+    for (let i = 0; i < sortedCpms.length; i++) {
+      if (i == 0) sortedCpms[i].bidding_cpm = Environment.ADS_MINIMUM_CPM;
+      else {
+        const bid = Math.min(
+          (i + 1) * Environment.ADS_MINIMUM_CPM,
+          sortedCpms[i].bidding_cpm,
+        );
+        sortedCpms[i].bidding_cpm =
+          Math.floor(bid / Environment.ADS_MINIMUM_CPM) *
+          Environment.ADS_MINIMUM_CPM;
+      }
+    }
+    return sortedCpms.reverse(); //most score show up first
```

Comment:
```suggestion
    return ads.sort((a,b) => releventScore[b....]-releventScore[a....]).map(....)
```

ควรใช้ sort desc by relevantScore ก่อนแล้วใช้ map เพื่อ return ค่านะครับ จะได้ลดจาก 4 loops เหลือ 2